### PR TITLE
Check if application can upgrade to the current channel

### PR DESCRIPTION
**Pull request by gabrielcocenza**
- generate_upgrade_plan can return empty plan
- create_upgrade_group can return empty plan
Closes: #123

### CODE REVIEW

**Review by aieri — DISMISSED** [review]

File: cou/apps/core.py [hunk 1]
```diff
@@ -422,7 +431,7 @@ def generate_upgrade_plan(self, target: str) -> UpgradeStep:
         for step in all_steps:
             if step:
                 upgrade_steps.add_step(step)
-        return upgrade_steps
+        return upgrade_steps if upgrade_steps.sub_steps else None
```
Comment (rgildein) [inline cou/apps/core.py:434]: What if we provide `empty` property, and basically skip any empty step instead of returning the `None`?

File: cou/steps/plan.py [hunk 2]
```diff
@@ -88,7 +90,7 @@ async def create_upgrade_group(
     target: str,
     description: str,
     filter_function: Callable[[OpenStackApplication], bool],
-) -> UpgradeStep:
+) -> Optional[UpgradeStep]:
```
Comment (rgildein) [inline cou/steps/plan.py:93]: I do not see how this can returns `None`. I think it will always returns `group_upgrade_plan`.

File: cou/steps/plan.py [hunk 3]
```diff
@@ -101,13 +103,14 @@ async def create_upgrade_group(
...
+    :rtype: Optional[UpgradeStep]
     """
     group_upgrade_plan = UpgradeStep(description=description, parallel=False)
     for app in filter(filter_function, apps):
         try:
-            app_upgrade_plan = app.generate_upgrade_plan(target)
-            group_upgrade_plan.add_step(app_upgrade_plan)
+            app_upgrade_plan: Optional[UpgradeStep] = app.generate_upgrade_plan(target)
+            if app_upgrade_plan:
+                group_upgrade_plan.add_step(app_upgrade_plan)
```
Comment (rgildein) [inline cou/steps/plan.py:113]: If we define empty property step we can do `if.app_upgrade_plan.empty`, which will be similar to what we have here. But what if we simply not show empty step in string representation and in execution we can skip any empty step. Like this we will have information about step, but we will know that's empty.

**Review by rgildein — COMMENTED** [review]
I do not like that `generate_upgrade_plan` can return `None` if there are no sub-steps, to me the plan should be simply empty and we can define `empty` property or batter will be to just define class representation better, so we can do `if step: ...`.

File: cou/apps/subordinate.py [hunk 4]
```diff
@@ -26,36 +25,36 @@
...
         """Post Upgrade planning.
 
         :param target: OpenStack release as target to upgrade.
         :type target: OpenStackRelease
         :return: Plan that will add post upgrade as sub steps.
-        :rtype: list[Optional[UpgradeStep]]
+        :rtype: list[UpgradeStep]
         """
-        return [None]
+        return [UpgradeStep()]
```
Comment (rgildein) [inline cou/apps/subordinate.py:57]: Why this cannot be simply empty list?
```suggestion
        return []
```

File: cou/steps/__init__.py [hunk 5]
```diff
@@ -134,6 +134,17 @@ def __repr__(self) -> str:
...
+    def __bool__(self) -> bool:
+        """Boolean magic method for UpgradeStep.
+
+        :return: True if there is at least one coroutine in a UpgradeStep
+        or in its sub steps.
+        :rtype: bool
+        """
+        if self._coro or any(step._coro for step in self.sub_steps):
+            return True
+        return any(bool(step) for step in self.sub_steps)
```
Comment (rgildein) [inline cou/steps/__init__.py:146]: I think that logic will be same if you use `and`, or not?
```suggestion
        return self._coro and any(bool(step) for step in self.sub_steps):
```

File: cou/steps/plan.py [hunk 6]
```diff
@@ -107,7 +107,8 @@ async def create_upgrade_group(
     for app in filter(filter_function, apps):
         try:
             app_upgrade_plan = app.generate_upgrade_plan(target)
-            group_upgrade_plan.add_step(app_upgrade_plan)
+            if app_upgrade_plan:
```
Comment (rgildein) [inline cou/steps/plan.py:110]: Is this if even needed? Empty steps will not be show to user and their `run` will do nothing.

File: tests/unit/steps/test_steps.py [hunk 7]
```diff
@@ -80,17 +80,65 @@ def test_step_eq(description, parallel, args):
...
     plan.sub_steps = [sub_step, UpgradeStep(description="a.b")]
 
+    assert str(plan) == ""
+
+
+def test_step_str_partially_show():
+    """Test UpgradeStep string representation when print UpgradeSteps that have coro."""
+    expected = "a\n\ta.a\n\t\ta.a.a\n"
+    plan = UpgradeStep(description="a")
+    sub_step = UpgradeStep(description="a.a")
```
Comment (rgildein) [inline tests/unit/steps/test_steps.py:135]: Can you also test to add empty step without even the description? I think we should skip such step in `__str__`, because right now, if I'm right, we will get `expected = "a\n\t\n\ta.a\n\t\ta.a.a\n"`

File: cou/steps/__init__.py (hunk 5 above)
Comment (gabrielcocenza) [reply cou/steps/__init__.py:146]: No. The Top plan doesn't have a coroutine (self._coro) but their sub steps will have.

Comment (rgildein) [reply cou/steps/__init__.py]: That is true, but `return self._coro is not None or any(bool(step) for step in self.sub_steps)` will work.

Comment (gabrielcocenza) [reply cou/steps/__init__.py]: That is what I did in the next commit

File: cou/apps/subordinate.py [hunk 8]
```diff
@@ -26,36 +25,36 @@
...
         """Post Upgrade planning.
 
         :param target: OpenStack release as target to upgrade.
         :type target: OpenStackRelease
         :return: Plan that will add post upgrade as sub steps.
-        :rtype: list[Optional[UpgradeStep]]
+        :rtype: list[UpgradeStep]
         """
-        return [None]
+        return []
```
Comment (aieri) [inline cou/apps/subordinate.py:57]: is this correct? If we claim to be returning an UpgradeStep should it not be `return [UpgradeStep()]`?

**Review by aieri — COMMENTED** [review]
Looks good overall, just a question

File: cou/apps/subordinate.py (hunk 8 above)
Comment (gabrielcocenza) [reply cou/apps/subordinate.py:57]: I was doing that one commit before and @rgildein suggested to change to an empty list. The type `list[UpgradeStep]` includes the empty list, So I think either way is fine.

**Review by aieri — APPROVED** [review]

**Review by rgildein — APPROVED** [review]
LGTM
